packages/cli/src/executions/execution.service.ts

```diff
@@ -505,23 +504,39 @@ export class ExecutionService {
 
 	public async annotate(
 		executionId: string,
-		updateData: { tags: string[]; vote: AnnotationVote | null },
+		updateData: ExecutionRequest.ExecutionUpdatePayload,
+		sharedWorkflowIds: string[],
 	) {
-		// FIXME: wrap in transaction
+		// Check if user can access the execution
+		const execution = await this.executionRepository.findIfAccessible(
+			executionId,
+			sharedWorkflowIds,
+		);
+
+		if (!execution) {
+			this.logger.info('Attempt to read execution was blocked due to insufficient permissions', {
+				executionId,
+			});
+
+			throw new NotFoundError('Execution not found');
+		}
+
+		// Create or update execution annotation
 		await this.executionAnnotationRepository.upsert(
 			{ execution: { id: executionId }, vote: updateData.vote },
 			['execution'],
 		);
 
-		if (updateData.tags) {
-			await this.annotationTagMappingRepository.overwriteTags(executionId, updateData.tags);
-		}
-
-		const execution = await this.executionRepository.findSingleExecution(executionId, {
-			includeAnnotation: true,
-			includeData: true,
+		// Upsert behavior differs for Postgres, MySQL and sqlite,
+		// so we need to fetch the annotation to get the ID
+		const annotation = await this.executionAnnotationRepository.findOneOrFail({
+			where: {
+				execution: { id: executionId },
+			},
 		});
 
-		return execution;
+		if (updateData.tags) {
+			await this.annotationTagMappingRepository.overwriteTags(annotation.id, updateData.tags);
+		}
```

Comment:
Ideally the `upsert` and `overwriteTags` should happen in the same transaction, but this is fine for now